src/_pytest/cacheprovider.py

```diff
@@ -131,7 +131,7 @@ def set(self, key, value):
             self.warn("cache could not write path {path}", path=path)
         else:
             with f:
-                json.dump(value, f, indent=2, sort_keys=True)
+                f.write(json.dumps(value, indent=2, sort_keys=True))
```

Comment:
can we move the dumps part over to above the opening of the file
simply to avoid corrupting the content in case a "dump" goes wrong

this doesn't have to be part of this pr, but it would be a nice to have

Reply:
Done.  Also added a changelog.